Multi-country targeting campaign - Adjust UI of campaign creation/edit and separate country selector when editing

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

This PR implemented the "Separate edit UI for both old and new campaigns" and "Change the rest of UI to fit with new design" parts of #1248:

- Separate the country(s) selector when editing old and new campaigns by
   - adding `multiple` prop to **\<AudienceSection\>**.
   - adding `allowMultiple` prop to **\<EditPaidAdsCampaignFormContent\>**.
- Display budget section in disabled status when no selected countries.
- Adjust wordings and style for the feature of multi-country targeting campaign.
- Workaround the problem that the tags' remove buttons are still displayed and workable when disabled **\<SelectControl\>** imported from `@woocommerce/components`.
   ![2022-02-23 13 54 55](https://user-images.githubusercontent.com/17420811/155268338-c7af7abe-b4ac-487c-a789-eff46d342ac9.png)
- Extra change: Prevent the shared onChange instance in tests of *audienceSection.test.js* - https://github.com/woocommerce/google-listings-and-ads/commit/33b16dc99d9c06c0eed7735d6261ae55c2e6e69d
   - The test would fail if exchange the order of tests in the "AudienceSection with multiple countries selector" group.

### Screenshots:

#### 📷  Edit single-country targeting campaign
![image](https://user-images.githubusercontent.com/17420811/155267640-47897496-59b6-4209-93ad-899fb3d3a768.png)

#### 📷  Edit multi-country targeting campaign
![image](https://user-images.githubusercontent.com/17420811/155267692-285404d5-657d-4558-8c08-434e72685685.png)

#### 📹  Disable budget section when no selected countries
https://user-images.githubusercontent.com/17420811/155268053-d34ce8ba-ca38-431b-b810-4ae51906db28.mp4

### Detailed test instructions:

💡  To facilitate testing, reviewers could simulate the `allowMultiple` property in the `receiveAdsCampaigns` function of [actions.js](https://github.com/woocommerce/google-listings-and-ads/blob/f085b54dc237c05cb82fb856ecbba9901dc5d52a/js/src/data/actions.js#L761-L766).
```js
// Add this before `return`
adsCampaigns = [
	{
		id: 1,
		name: 'Single targeting campaign',
		status: 'enabled',
		amount: 123456,
		country: 'US',
		allowMultiple: false,
	},
	{
		id: 2,
		name: 'Multiple targeting campaign',
		status: 'enabled',
		amount: 654321,
		country: 'JP',
		allowMultiple: true,
	},
];
```

#### Edit campaign

1. Go to edit campaigns for both single- and multi-country targeting campaigns.
2. The UI should be close to the visual design.

#### Create campaign

1. Go to create a new campaign
2. The UI should be close to the visual design.
3. Before selecting countries, the budget section should be displayed in disabled style.
4. After selecting one or more countries, the budget section should be displayed in enabled style.
5. After clearing all selected countries, the budget section should be back to be displayed in disabled style.

### Changelog entry

eason9487 committed:
Move edit-paid-ads-campaign-form-content.js to shared components folder

eason9487 committed:
Add `multiple` prop to AudienceSection for selecting countries and adjust style

eason9487 committed:
Prevent the shared onChange instance in tests of audienceSection.test.js

eason9487 committed:
Add `allowMultiple` prop to EditPaidAdsCampaignFormContent for supporting both campaign targeting features

eason9487 committed:
Adjust wordings for multi-country targeting campaign

eason9487 committed:
Adjust style for multi-country targeting campaign

eason9487 committed:
Display budget section in disabled status when no selected countries

eason9487 committed:
Workaround the problem that the tags' remove buttons are still displayed and workable when disabled SelectControl

ecgan commented on js/src/components/paid-ads/budget-section/index.js:
```diff
@@ -49,6 +76,8 @@ const BudgetSection = ( props ) => {
 								suffix={ currency }
 								{ ...getInputProps( 'amount' ) }
+								value={ amountValue }
+								readOnly={ disabled }
```
❔  I think this should be `disabled` instead of `readonly`?

```suggestion
								disabled={ disabled }
```

Here's my testing video with the current `readOnly={ disabled }` code:

https://user-images.githubusercontent.com/417342/155371525-cea36590-24c6-4268-b8ba-bbc3d7004515.mp4

eason9487 replied:
In order to get as close to the visual design while not overriding the style of wp-components, using `readOnly` is a compromise.

I [added another workaround style](https://github.com/woocommerce/google-listings-and-ads/pull/1268/commits/17c766ab83955f84344a34b36f80398f7f1fba88) to avoid the focused style though it still has to override them 😅 , I think it'd be better than overriding the disabled style.

eason9487 replied:
> I [added another workaround style](https://github.com/woocommerce/google-listings-and-ads/pull/1268/commits/17c766ab83955f84344a34b36f80398f7f1fba88) to avoid the focused style [...]

I [committed another one](https://github.com/woocommerce/google-listings-and-ads/pull/1268/commits/2b07743f4242b90da99b263ce67c970218b2fd51) to fix that focused style is still appears when clicking on its `<label>`. It should be ok now 😅 

![Kapture 2022-02-24 at 14 44 08](https://user-images.githubusercontent.com/17420811/155472443-2c42e9a3-53f9-4ca6-8b96-b143e575dde3.gif)

ecgan replied:
> In order to get as close to the visual design while not overriding the style of wp-components, using `readOnly` is a compromise.

💅  Personally sometimes I would use my own best judgement, and in this case, I would just use `disabled` to achieve the goal of making it not editable, even though it may be slightly off in visual design. I get your point here though, and I'm fine with that, and thank you for your desire on getting close to visual design. 🙂  👍

ecgan commented on js/src/wcdl/select-control/index.scss:
```diff
@@ -44,5 +44,20 @@
 			top: unset;
 		}
+
+		// Workaround the problem that the tags' remove buttons are still displayed when disabled,
+		// no matter the `inlineTags` are true or false.
+		.is-disabled {
+			.woocommerce-select-control__tags,
+			+ .woocommerce-select-control__tags {
+				.woocommerce-tag__remove {
+					display: none;
+				}
+				.woocommerce-tag__text {
+					padding: 0 $grid-unit-10;
+					border-radius: 12px;
+				}
+			}
+		}
```
👍  Good stuff hiding the tag's remove button with CSS! 😃

ecgan commented on js/src/components/paid-ads/audience-section.js:
```diff
@@ -2,25 +2,71 @@
+	const countryNameMap = useCountryKeyNameMap();
 	const inputProps = getInputProps( 'country' );
+
+	/**
+	 * Hack to prevent tags being removed even when disabled.
+	 * This could be removed after replacing <AudienceCountrySelect> with <TreeSelectControl>.
+	 */
 	if ( disabled ) {
 		inputProps.onChange = undefined;
 	}
```
❔  Since we hide the tag's remove button via SelectControl's CSS, I think we actually don't need the code here. I tested commented out the code and it works for me.

```suggestion
```

eason9487 replied:
I had [removed this](https://github.com/woocommerce/google-listings-and-ads/commit/26ac0d37e9add0f9dfa8a1de8e9d9f700c924acc#diff-5c6f0d2e75d046351c001bcf2b5de4756de5193bae214db825e21d17b9597c05L20-L22) while developing. And then I recalled `@woocommerce/components` is a runtime version injected via DEWP. I worried it might break again someday if only has the CSS workaround, so I added this back to have a second layer of defense to secure the wanted disabled function.

I will open an issue to upstream repo. Hope we can remove this workaround soon.

ecgan replied:
That's a careful thinking and a totally valid point! Agree 😃 👍

ecgan commented on js/src/components/paid-ads/audience-section.js:
```diff
@@ -2,25 +2,71 @@
+	const selector = multiple ? (
+		<AudienceCountrySelect
+			label={ __( 'Select countries', 'google-listings-and-ads' ) }
+			helperText={ countrySelectHelperText }
```
💅  ❔  Related to `countrySelectHelperText`, we don't need it as props, we can have the text inline here, like: 

```suggestion
			helperText={ __(
				'Once a campaign has been created, you cannot change the target country(s).',
				'google-listings-and-ads'
		  ) }
```

eason9487 replied:
`<AudienceSection>` is also used in campaign creation and this helper is only shown when editing. So I thought it should be a bit better rather than passing a `@param {'create'|'edit'} mode` for `<AudienceSection>` to control this.

ecgan replied:
> `<AudienceSection>` is also used in campaign creation and this helper is only shown when editing.

D'oh, somehow I missed that. 🤦 

So this is my thinking now:

- Campaign create: section is enabled, there is no helper text.
- Campaign edit: section is disabled, we show helper text to explain that they can't change the country.

From the above, the helper text could depend on the `disabled` prop and we don't need `countrySelectHelperText` prop? 🤔 

```suggestion
			helperText={
				disabled &&
				__(
					'Once a campaign has been created, you cannot change the target country(s).',
					'google-listings-and-ads'
				)
			}
```

eason9487 replied:
Thanks for the suggestion! This does give the expected results without adding additional prop. 👍 This does give the expected results without an additional prop.

However, when determined by `disabled`, this implicitly equates disabled logically with creation and edit. While the current result will be correct, it also implicitly reduces maintainability, e.g. if later need to disable the selector control when creating, then `disabled` may not be applicable for determining the helper.

On the other hand, when TreeSelectControl is integrated later, there is still a sentence similar to helper text that needs to be displayed. Looks like they both are a kind of helper text used in the same meaning though the positions are different. The `countrySelectHelperText` may still be used later, depending on the design of TreeSelectControl and subsequent integration considerations.

![image](https://user-images.githubusercontent.com/17420811/155642917-47b84c40-e3b3-4c46-9f1c-3954d78c89c2.png)

ecgan replied:
Alright, let's go with what we have.

ecgan commented on js/src/components/paid-ads/audience-section.js:
```diff
@@ -2,25 +2,71 @@
+	) : (
+		<SelectControl
+			label={ __( 'Select one country', 'google-listings-and-ads' ) }
+			help={ countrySelectHelperText }
```
💅  ❔  Related to `countrySelectHelperText`, same thing, we don't need it as props, we can have the text inline here, like: 

```suggestion
			helperText={ __(
				'Once a campaign has been created, you cannot change the target country.',
				'google-listings-and-ads'
		  ) }
```

eason9487 replied:
Same as https://github.com/woocommerce/google-listings-and-ads/pull/1268#discussion_r813547502

ecgan replied:
Similar to https://github.com/woocommerce/google-listings-and-ads/pull/1268#discussion_r813954523 😆 

```suggestion
			helperText={
				disabled &&
				__(
					'Once a campaign has been created, you cannot change the target country.',
					'google-listings-and-ads'
				)
			}
```

ecgan commented on js/src/wcdl/section/index.js:
```diff
@@ -1,14 +1,31 @@
+ * @param {JSX.Element} [props.children] Section content at the right side.
+ * @param {boolean} [props.disabled] Whether display the whole section in disabled style.
+ */
+const Section = ( { title, description, children, disabled } ) => {
+	const className = classnames(
+		'wcdl-section',
+		disabled ? 'wcdl-section--is-disabled' : false
+	);
```
💅  Maybe also consider allowing `Section` consumer to specify their own className?

```suggestion
const Section = ( { title, description, children, disabled, className } ) => {
	const className = classnames(
		'wcdl-section',
		disabled ? 'wcdl-section--is-disabled' : false,
		className
	);
```

eason9487 replied:
It would have some eslint errors by committing with the suggestion codes. I added in https://github.com/woocommerce/google-listings-and-ads/pull/1268/commits/b93ee757531a6f339b96d267afa52d82365e65b9

ecgan replied:
Nice! No worry about not using the suggested change code, it is understandable and I don't mind that at all, the more important thing is to communicate the idea. Thanks for the commit! 😃 👍

ecgan commented on js/src/pages/edit-paid-ads-campaign/edit-paid-ads-campaign-form.js:
```diff
@@ -82,6 +82,7 @@ const EditPaidAdsCampaignForm = ( props ) => {
 						/>
 						<EditPaidAdsCampaignFormContent
 							formProps={ formProps }
+							allowMultiple={ campaign.allowMultiple }
```
💅  ❔ Instead of passing `allowMultiple` / `multiple` down the component tree, what do you think if we put the `campaign.multiple` into form value, and the child components can just pull out `multiple` value from formProps? e.g.:

```
		<Form
			initialValues={ {
				id: campaign.id,
				amount: campaign.amount,
				country: [ campaign.country ],
				multiple: campaign.multiple // put `multiple` into form value.
			} }

eason9487 replied:
Not sure if it's a good idea.

Since the coexisting of "old" and "new" campaigns is a special logic. I planned to put all relevant logic at two places, 1) here the `allowMultiple` prop and 2) adding `allowMultiple` property to each campaign entity when receiving API data in actions.js.

That's said, this special logic is isolated from other places. And the `multiple` is designed as a normal prop for `<AudienceSection>` to indicate the selector type, but it doesn't need to know the special logic.

Another thought is, the purpose of `initialValues` seems to be more towards defining form data, and if passing down other props directly is not an obvious bad smell code, I'd suggest we should try to avoid passing prop along with form `initialValues`.

ecgan replied:
> adding `allowMultiple` property to each campaign entity when receiving API data in actions.js.

Ahh, thanks for this, I didn't know that. I thought `allowMultiple` is part of the campaign object returned from the API.

Just want to throw out an idea that I thought about (and related to the `targeted_locations` comment in https://github.com/woocommerce/google-listings-and-ads/pull/1268#issuecomment-1049508806). Since the form is storing the campaign values, we could probably extract the formProps value and determine whether it is a new multi-country campaign or old single-country campaign. Here's a simple illustration of the idea:

```js
// Form.
<Form
	  initialValues={ {
		    id: campaign.id,
		    amount: campaign.amount,
		    country: [ campaign.country ],
		    targeted_locations: campaign.targeted_locations, // NEW: I think we would need this eventually?
	  } }
>...

// Function containing the logic to determine whether it is a new multiple campaign, based on targeted_locations.
const isCampaignMultiple = (campaign) => !! campaign.targeted_locations

// Within child component.
const isMultiple = isCampaignMultiple(formProps.values) // values is essentially campaign here.
```

I'm not sure if the above is a good idea, but I guess it is something that you can evaluate and find out. 😄 

Anyway, with your explanation above, I think we are good with what we have now, we can go with it. 🙂  👍

eason9487 replied:
Thanks for the idea! I tried a different way when integrating `targeted_locations` data. Will still use the form value `country`, renamed to `countryCodes`, for showing the single country or multiple countries. And that country codes will be prepared by an adapter in actions.js. It looks like:

```js
// In actions.js
function adapteAdsCampaign( campaign ) {
	const allowMultiple = campaign.targeted_locations.length > 0;
	const displayCountries = allowMultiple
		? campaign.targeted_locations
		: [ campaign.country ];
	return {
		...campaign,
		allowMultiple,
		displayCountries,
	};
}

// Form
<Form
	initialValues={ { amount, countryCodes: displayCountries } }
	validate={ handleValidate }
	onSubmit={ handleSubmit }
>
```

Sorry I should have added the relevant design in my head to the context of this PR. 🙏  This approach will be more clearly outlined in the subsequent PR

ecgan replied:
No worry, and thanks for the explanation, we will work through it in the subsequent PR then. 🙂

ecgan commented on js/src/components/paid-ads/budget-section/index.js:
```diff
@@ -13,30 +13,57 @@ import BudgetRecommendation from './budget-recommendation';
 	const { googleAdsAccount } = useGoogleAdsAccount();
-	const monthlyMaxEstimated = getMonthlyMaxEstimated( values.amount );
+
+	const amountValue = disabled ? '' : values.amount;
+	const monthlyMaxEstimated = disabled
+		? ''
+		: getMonthlyMaxEstimated( values.amount );
```
❓  So here the form value `amount` doesn't change, and we conditionally display things based on `disabled`. 

I'm not sure if this might be a problem or not. When users clear the list of countries and then type a new country, the previous daily average cost value reappear. See a demo video below:

https://user-images.githubusercontent.com/417342/155393375-4e8390a0-e9c7-4204-824b-85fcf2a00621.mov

I'm thinking it might be better if we update the form value `amount` to e.g. `undefined` when the countries are removed? If we do that, then I think we don't need to have these 4 lines of code. What do you think? 🤔

eason9487 replied:
I think it depends on which scenario is the priority from the UI design/user perspective.

- As a user, I expect the form content I filled to be cleared only when I actively reset it.
- As a user, I expect when I fill a form, if an important field changes, the related important fields must be reset, so that I can be prompted to rethink and confirm that those field values are correct by filling them again.

But I can't tell which one or any others is exactly the priority. 😅 

---

Hey @j111q 👋 , could you help with inputting which is the preferred one or other options?

j111q replied:
Hmmm yeah thanks for the question and agree with the two competing priorities you stated Eason! 🤔 

I think for this one, it's cleaner to prioritise (2): `As a user, I expect when I fill a form, if an important field changes, the related important fields must be reset, so that I can be prompted to rethink and confirm that those field values are correct by filling them again.`

I don't anticipate many users removing all their countries after they've already filled in the amount -- but if that does happen, let's reset the form for them, and make it clear that they need to fill in the budget again just in case it has changed. 👍

eason9487 replied:
> I think for this one, it's cleaner to prioritise (2): `As a user, I expect when I fill a form, if an important field changes, the related important fields must be reset, so that I can be prompted to rethink and confirm that those field values are correct by filling them again.`


@j111q thanks for replying!

@ecgan, I updated the behavior to this scenario in https://github.com/woocommerce/google-listings-and-ads/pull/1268/commits/e07c7edac700e0161dfbbc806839841771d7bca7. Could you help with second round review?

![Kapture 2022-02-24 at 15 26 24](https://user-images.githubusercontent.com/17420811/155477956-e4dbcd8a-4a51-4fe4-8f6d-0dd6ad6a442b.gif)

ecgan replied:
@eason9487 , I reviewed the code and left my comment in https://github.com/woocommerce/google-listings-and-ads/pull/1268#discussion_r814090894.

ecgan reviewed (approved):
Tested as per test instruction, things seem to be working as expected. Ran `npm run test:js`, all pass.

❓ One question that I have in my mind while testing with the test instruction and test data is about the multi-country targeting campaign. Instead of `country: 'JP',`, shouldn't it be like `country: ['JP', 'US],` since it is multi-country targeting? 🤔 

I left some comments below, most are slight code improvement in my opinion. There is one ❓  comment below about `amount` form value when users change the countries, which might or might not be a problem.

Since there are no major issues and things are working as expected, I'm approving this.

eason9487 committed:
Prevent the read-only input from having the focused border style.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1268#discussion_r813123513

eason9487 committed:
Allow `<Section>` to be able to append additional class name.

Co-authored-by: Gan Eng Chin <ecgan@users.noreply.github.com>

Address https://github.com/woocommerce/google-listings-and-ads/pull/1268#discussion_r813230882

eason9487 commented:
@ecgan

> ❓ One question that I have in my mind while testing with the test instruction and test data is about the multi-country targeting campaign. Instead of `country: 'JP',`, shouldn't it be like `country: ['JP', 'US],` since it is multi-country targeting?

Thanks for the review comments! 

After backend API is finished, the campaign entity will be the following structure.

> It will be an old campaign if the `targeted_locations` is an empty array.

```json
[
  {
    "id": 16165825406,
    "name": "Campaign 2022-02-09 09:31:52",
    "status": "enabled",
    "amount": 50,
    "country": "TW",
    "targeted_locations": []
  },
  {
    "id": 16178695424,
    "name": "Campaign 2022-02-10 06:29:08",
    "status": "enabled",
    "amount": 245,
    "country": "TW",
    "targeted_locations": ["TW", "GB"]
  },
]
```

And the Smart Shopping Campaigns can have an **immutable single** `sales_country` when creating.

https://github.com/woocommerce/google-listings-and-ads/blob/889234ada862a315a5f2aa3d592450e969f0c316/src/API/Google/AdsCampaign.php#L117-L122

https://github.com/woocommerce/google-listings-and-ads/blob/889234ada862a315a5f2aa3d592450e969f0c316/src/API/Google/AdsCampaign.php#L292-L297

After implementing the multi-country targeting campaign, the `sales_country` will be changed to take the country from WC Settings when creating, so no matter the "new" or "old" campaign, they will always have the single value `country` property.

eason9487 committed:
Fix that the `<input>` is still being triggered focused style when clicking on its `<label>`.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1268#discussion_r813123513

eason9487 committed:
Reset the filled amount when `disabled` changes to true in BudgetSection.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1268/files#r813246789

ecgan commented on js/src/components/paid-ads/budget-section/index.js:
```diff
@@ -13,30 +14,70 @@ import BudgetRecommendation from './budget-recommendation';
 	// Display the currency code that will be used by Google Ads, but still use the store's currency formatting settings.
 	const currency = googleAdsAccount?.currency;
 
+	/**
+	 * In addition to the initial value setting during initialization, when `disabled` changes
+	 * - from false to true, then clear filled amount to `undefined` for showing a blank <input>.
+	 * - from true to false, then reset amount to the initial value passed from the consumer side.
+	 */
+	const initialAmountRef = useRef( amount );
+	useEffect( () => {
+		const nextAmount = disabled ? undefined : initialAmountRef.current;
+		setValue( 'amount', nextAmount );
+	}, [ disabled, setValue ] );
```
❓ Interesting. There are two things I like to touch on: 

The first one is about `initialAmountRef` and `nextAmount`. They are used to "reset amount to the initial value passed from the consumer side". When I read https://github.com/woocommerce/google-listings-and-ads/pull/1268#discussion_r813246789 (with @eason9487 and @j111q), my interpretation of "reset" is to make the value undefined (make it blank), and one reason for it is that when it is disabled, the input box is blank. If we were to reset it to initial value (i.e. enabled with initial value "5", then disabled with blank value, then enabled with initial value "5" again), the disappearing and re-appearing of value feel weird for me. 😅 

The second one is about `useEffect`. Instead of using `useEffect`, my thought is to tap into the SelectControl's `onChange` event handler inside `AudienceSection` component, i.e. in `onChange`, if the new value is empty array (no country), then we call `setValue( 'amount', undefined )`. What do you think? 🤔

eason9487 replied:
> If we were to reset it to initial value (i.e. enabled with initial value "5", then disabled with blank value, then enabled with initial value "5" again), the disappearing and re-appearing of value feel weird for me.

It still depends on scenarios. Usually, forms are set to initial/default values when "reset" and set to blank/default values when "clear". And from an input component design perspective, it should use the given initial value. Otherwise, as a developer, it looks like a bug if the initial value on \<Form\> is `5` and the input shows blank rather than the given initial one when changing from disabled to enabled.

> The second one is about `useEffect`. Instead of using `useEffect`, my thought is to tap into the SelectControl's `onChange` event handler inside `AudienceSection` component, i.e. in `onChange`, if the new value is empty array (no country), then we call `setValue( 'amount', undefined )`. What do you think? 🤔

Yes, it's better to reset amount relying on the changing of country values, but resetting amount in country selector will increase coupling between these components. I don't think it's a good idea.

ecgan replied:
> resetting amount in country selector will increase coupling between these components

Yeah, I think I get what you mean, I wish there is a way to put that logic in the form component level and not in the `AudienceSection`, but I don't think there is one.

I guess we can go with what we have here.

ecgan commented on js/src/components/app-input-control/index.scss:
```diff
@@ -10,4 +10,10 @@
 		}
 	}
+
+	// Hack to prevent the read-only input from having the focused border style.
+	// Ref: https://github.com/WordPress/gutenberg/blob/%40wordpress/components%4012.0.8/packages/components/src/input-control/styles/input-control-styles.js#L238-L247
+	&--no-pointer-events {
+		pointer-events: none;
+	}
```
❔ Since it is a "Hack to prevent the _read-only input_ from having the focused border style", I'm thinking it may be better to name the class like `&--read-only`? 

Side note: I don't think [`:read-only`](https://developer.mozilla.org/en-US/docs/Web/CSS/:read-only) pseudo-class would work here (since the CSS class applies to div), right?

eason9487 replied:
This class is exactly applying `pointer-events: none;` style. Not sure why it's better to name with `&--read-only`?

I rephrased the code comments in https://github.com/woocommerce/google-listings-and-ads/pull/1268/commits/e56653d518494591231d47562f7e1cc18b43566b. Hope the intention could be expressed more clearly.

> Side note: I don't think [`:read-only`](https://developer.mozilla.org/en-US/docs/Web/CSS/:read-only) pseudo-class would work here (since the CSS class applies to div), right?

Right, a read-only DOM is still focusable.

ecgan replied:
> Not sure why it's better to name with `&--read-only`?

The reason is to make the CSS class describe what it is, instead of how it is supposed to look like. For example, if we want a read-only element to have no pointer events, display text in yellow color and with red background, the class naming "read-only" would be better than "no-pointer-events".

ecgan commented on js/src/components/app-input-control/index.js:
```diff
@@ -10,11 +10,26 @@ import { __experimentalInputControl as InputControl } from '@wordpress/component
+	const { className, noPointerEvents, ...rest } = props;
+	const wrapperClassName = classnames(
+		baseClassName,
+		noPointerEvents && `${ baseClassName }--no-pointer-events`,
```
❔ I think we don't need `noPointerEvents`, we can just rely on `readOnly` props? 🤔

eason9487 replied:
Not sure your meaning. Relying on `readOnly ` will still have the problem you reviewed in  https://github.com/woocommerce/google-listings-and-ads/pull/1268#discussion_r813123513.

ecgan replied:
Sorry, I should have been clearer that this is related to https://github.com/woocommerce/google-listings-and-ads/pull/1268#discussion_r814105386. I hope the following illustrates better, we use the `readOnly` prop, add the "read-only" CSS class, and for the "read-only" class, we say no pointer events: 

```suggestion
		readOnly && `${ baseClassName }--read-only`,
```

And then in the CSS: 

```
	// For read-only elements, we don't want pointer events.
	// (And the read-only element is open to other CSS modifications like color, text size etc.
	&--read-only {
		pointer-events: none;
	}
```

eason9487 replied:
Thanks for the explanation!

The following reply also includes replies to [another comment](https://github.com/woocommerce/google-listings-and-ads/pull/1268#discussion_r815028101).

```js
readOnly && `${ baseClassName }--read-only`
```

I don't think it's a good idea to force binding the `readonly` attribute and `pointer-events: none` style inside an elementary component. This will decrease the flexibility of its use.

> For example, if we want a read-only element to have no pointer events, display text in yellow color and with red background, the class naming "read-only" would be better than "no-pointer-events".

In addition, the naming of `--read-only` also brings an ambiguous meaning because they are different.
- A **\<input\>** with `readonly` attribute still works with pointer events.
- A **\<input\>** with `pointer-events: none` style is still editable.

if there're other styles should be applied when setting `pointer-events: none` on elements, then I would still name it `&--no-pointer-events`. And class having other styles along with `readonly` attribute would be `&--read-only`, `&:read-only` or `&[readonly]` then.

> The reason is to make the CSS class describe what it is, instead of how it is supposed to look like.

`pointer-events: none` doesn't change the appearance of \<InputSelect\>, it changes the way it interacts, it's a style but acts more like an attribute. I don't understand why you think that the naming is closer to "how it is supposed to look like". If want a more precise description of what it is, `--no-pointer-events` should be a precise description already. Or what about `--ignore-pointer-events` or `--disable-pointer-events` instead? Or do you have an ideal naming without above the ambiguity?

ecgan replied:
>> The reason is to make the CSS class describe what it is, instead of how it is supposed to look like.

> I don't understand why you think that the naming is closer to "how it is supposed to look like".

I'll slightly rephrase my original sentence to this: "The reason is to make the CSS class describe what it is semantically (that is generally understandable, by a developer or even a user), instead of how it is supposed to look like or act like in technical terms."

For example, a div containing error message would have class like ".error", warning message with ".warning", dangerous destructive button with ".danger". Those CSS class describe what they are semantically. Then we would have technical specific styling (text color, background color, border etc) for those CSS class in CSS files. 

> In addition, the naming of `--read-only` also brings an ambiguous meaning because they are different.
> 
> * A **<input>** with `readonly` attribute still works with pointer events.
> * A **<input>** with `pointer-events: none` style is still editable.

The way I see "read-only" here is not in the technical sense but in the semantical sense. When we say the term "read only" (like "error", "warning", "disabled" etc), likely everyone can understand the meaning of it. However, we can't really say the same with "no-pointer-events".

> if there're other styles should be applied when setting `pointer-events: none` on elements, then I would still name it `&--no-pointer-events`.

I take it that you mean it like this?

```css
.no-pointer-events { /* I simplify the class name here. */
	pointer-events: none;

	/* other styles... */
	color: red;
	background-color: pink;
	border: solid 1px red;
	/* ... and possibly many other styles. */
}
```

If that's the case, I find it difficult / weird to wrap my head around it, e.g. a `<div class="no-pointer-events">` with those UI styles applied to it kinda goes against my intuition. 😅 

> Or what about `--ignore-pointer-events` or `--disable-pointer-events` instead?

Those aren't very much different from `no-pointer-events`. I see "pointer events" as a technical implementation term and not semantic term. It is like `<div class="green-text black-background disable-pointer-events">`, with class names indicating implementation; I would prefer it like `<div class="matrixbox">` and have the technical implementation inside `.matrixbox { ... }` CSS.

---

I hope I get my points across. If you prefer your current implementation over my suggestions comments above, feel free to go ahead and merge, it is not a blocking issue and I don't want to hold up the project on this. 😄 🙏

eason9487 replied:
> [...], instead of how it is supposed to look like or act like in technical terms."

Sorry, I don't get why a code or a name in codes for readers, who are mostly devs, is not supposed to look like a technical term? I think it depends more on the abstraction level or readability (under a lesser ambiguity).

> For example, a div containing error message would have class like ".error", warning message with ".warning", dangerous destructive button with ".danger". Those CSS class describe what they are semantically. Then we would have technical specific styling (text color, background color, border etc) for those CSS class in CSS files.

I agree with this. And even the `&--no-pointer-events` is exactly fit with the semantical need because `pointer-events: none` only affect the behavior and doesn't change any appearance of \<InputSelect\> nor make something looks like read-only.

> If that's the case, I find it difficult / weird to wrap my head around it, e.g. a `<div class="no-pointer-events">` with those UI styles applied to it kinda goes against my intuition.

If that makes it difficult to understand, then I will use `&--no-pointer-events-with-what-it-is-semantically-style` such as `&--no-pointer-events-with-disabled-style`. Or separate it to `&--no-pointer-events` and `&--disabled-entire-selector`.

> I would prefer it like `<div class="matrixbox">` and have the technical implementation inside `.matrixbox { ... }` CSS.

😕  Sorry, I really don't get why is matrixbox. When seeing a `.matrixbox` CSS class, how could it be easily and accurately deduced the class will disable point events?

---

> I hope I get my points across. If you prefer your current implementation over my suggestions comments above, feel free to go ahead and merge, it is not a blocking issue and I don't want to hold up the project on this.

Thanks for the explanation!

But sorry I still feel confused. 🙏 
I will proceed with merging. If you think this implementation is not applicable from some perspectives, I'd be happy to discuss it separately later, or have you improve it in another PR. 🙇‍♂️

ecgan reviewed (commented):
Tested, things working okay. I left one ❓ comment about the "reset amount" code to get more clarification and understanding, and two ❔ comments about namings.

eason9487 committed:
Rephrase the code comments in the app-input-control/index.scss

Address https://github.com/woocommerce/google-listings-and-ads/pull/1268#discussion_r814105386

[skip ci] - only change code comments
